exonum/src/node/state.rs

```diff
@@ -856,23 +808,37 @@ impl State {
     }
 
     /// Adds propose from other node. Returns `ProposeState` if it is a new propose.
-    pub fn add_propose(&mut self, msg: &Propose) -> Option<&ProposeState> {
+    pub fn add_propose(
+        &mut self,
+        msg: &Propose,
+        transactions: &MapIndex<&&Snapshot, Hash, RawMessage>,
+        transaction_pool: &KeySetIndex<&&Snapshot, Hash>,
+    ) -> Result<&ProposeState, failure::Error> {
         let propose_hash = msg.hash();
-        let txs = &self.transactions.read().expect("Expected read lock");
         match self.proposes.entry(propose_hash) {
-            Entry::Occupied(..) => None,
+            Entry::Occupied(..) => bail!("Propose already found"),
             Entry::Vacant(e) => {
-                let unknown_txs = msg.transactions()
-                    .iter()
-                    .filter(|tx| !txs.contains_key(tx))
-                    .cloned()
-                    .collect::<HashSet<Hash>>();
+                let mut unknown_txs = HashSet::new();
+                for hash in msg.transactions() {
+                    if transactions.get(hash).is_some() {
+                        if !transaction_pool.contains(hash) {
```

Comment:
let's merge those two `if`s with `||`, maybe? I don't like the amount of nested scopes

Reply:
I cant use `or` there because of three possible cases:
- Transaction was found in database but not found in pool - this transaction already commited and propose of this transaction is invalid.
- Transaction was found in database and it is in pool - this transaction is valid for new block.
- Transaction was not found in pool - this transaction is unknown, we need to request it later.